Src/IronPython.Modules/_codecs.cs

```diff
@@ -406,63 +406,104 @@ public static PythonTuple utf_7_encode(CodeContext context, string input, string
 
         #region Utf-32 Functions
 
-        public static PythonTuple utf_32_decode(CodeContext context, [BytesConversion]IList<byte> input, string errors = "strict", bool final = false)
-            => DoDecode(context, "utf-32", Encoding.UTF32, input, errors, final).ToPythonTuple();
+        public static PythonTuple utf_32_decode(CodeContext context, [BytesConversion]IList<byte> input, string errors = "strict", bool final = false) {
+            PythonTuple res = utf_32_ex_decode(context, input, errors, 0, final);
+            return PythonTuple.MakeTuple(res[0], res[1]);
+        }
 
         public static PythonTuple utf_32_encode(CodeContext context, string input, string errors = "strict")
-            => DoEncode(context, "utf-32", Encoding.UTF32, input, errors, true).ToPythonTuple();
+            => DoEncode(context, "utf-32", Utf32LeBomEncoding, input, errors, includePreamble: true).ToPythonTuple();
 
-        #endregion
+        public static PythonTuple utf_32_ex_decode(CodeContext context, [BytesConversion]IList<byte> input, string errors = "strict", int byteorder = 0, bool final = false) {
 
-        public static PythonTuple utf_32_ex_decode(CodeContext context, [BytesConversion]IList<byte> input, string errors = "strict")
-            => utf_32_ex_decode(context, input, errors, null, null);
+            Tuple<string, int> res;
 
-        public static PythonTuple utf_32_ex_decode(CodeContext context, [BytesConversion]IList<byte> input, string errors, object byteorder, object final) {
-            byte[] lePre = Encoding.UTF32.GetPreamble();
+            if (byteorder != 0) {
+                res = (byteorder > 0) ?
+                    DoDecode(context, "utf-32-be", Utf32BeEncoding, input, errors, final)
+                :
+                    DoDecode(context, "utf-32-le", Utf32LeEncoding, input, errors, final);
 
-            string instr = Converter.ConvertToString(input);
-            bool match = true;
-            if (instr.Length > lePre.Length) {
+            } else {
+                byteorder = Utf32DetectByteorder(input);
+                res = (byteorder > 0) ?
+                    DoDecode(context, "utf-32-be", Utf32BeBomEncoding, input, errors, final)
+                :
+                    DoDecode(context, "utf-32-le", Utf32LeBomEncoding, input, errors, final);
+            }
+
+            return PythonTuple.MakeTuple(res.Item1, res.Item2, byteorder);
+        }
+
+        private static int Utf32DetectByteorder(IList<byte> input) {
+            byte[] lePre = BOM_UTF32_LE;
```

Comment:
Hmm, I had assumed that this was correct since it matched what we had before, but it doesn't match CPython when `input == BOM`.
```Python
assert _codecs.utf_32_ex_decode(b"\x00\x00\xfe\xff") == ('', 4, 1)
```

I think a nicer implementation would be:
```C#
        private static int Utf32DetectByteorder(IList<byte> input) {
            if (input.StartsWith(BOM_UTF32_LE)) return -1;
            if (input.StartsWith(BOM_UTF32_BE)) return 1;
            return 0;
        }
```

Reply:
Good catch. I will add a test for this case.

The condition '>' operator should have been '>='. But I like your succinct version better. I didin't know we had `IList<byte>.StartsWith`. BTW, it can also be applied in `StringOps.GetStartingOffset`.

I'm going to read `IListOfByteOps` to see what other goodies we have around.